Add support for packetfilter API and iptables driver

yboaron opened the pull request:
To allow smooth transition to nftables a new packetfilter component is introduced, packetfilter 
provides an API to create chains, and rules in a generic way and should 
include pluggable drivers for iptables and nftables.

Packetfilter also includes a static function that defines the driver to use.

This PR includes:
1. Refactoring the current iptables code to be used as pluggable driver.
2. Fake driver implementation for unit tests.
3. Update clients to use packetfilter APIs

Generic named set API will be pushed in a separate PR. 

check [1] for nftables support high level design .

Co-authored-by: Tom Pantelis <tompantelis@gmail.com>

[1] 
https://docs.google.com/document/d/1PAjU61XUGaQ2qZZu_66clxadC997lsBGYcjydYEayR0/edit?usp=sharing

tpantelis commented on pkg/packetfilter/types/types.go:
```diff
@@ -0,0 +1,160 @@
+limitations under the License.
+*/
+
+package types
```
I don't think it's necessary to define types in a sub-package and it's not really a pattern I've seen. Consumers would reference, eg, `types.ChainIPHook` which isn't clear as opposed to `packetfilter.ChainIPHook` (unless consumers rename the import which is kind of inconvenient).

yboaron replied:
Agree that defining the types under packetfilter package is more appropriate here (and that also was my original code as well). But this created a package cyclic dependency issue.

1. Client imports packetfilter
2. packetfilter imports iptables (and nftables in the future) - to include these packages in compiled code
3. iptables (and nftables) should use the types/structures defined in packetfilter

Any idea how to have the types under packetfilter and avoid the cyclic dependency?

Maybe we can :
 - underscore import  iptables (and nftables) from the main package to include these packages in compiled code
 - client imports packetfilter 
 - iptables and nftables import packetfilter

tpantelis replied:
We could eliminate number 2 above and have _main.go_ explicitly import the implementation packages and register them (or make registration implicit via an `init` function in the package).

yboaron replied:
OK, I updated the code

tpantelis commented on pkg/packetfilter/types/types.go:
```diff
@@ -0,0 +1,160 @@
+	DeleteGlobalEgressForNamespace(handle GlobalEgressHandle) error
+	AddPodIPToGlobalEgress(handle GlobalEgressHandle, podIP string) error
+	RemovePodIPToGlobalEgress(handle GlobalEgressHandle, podIP string) error
+}
```
This interface seems a bit broad and perhaps disparate, ranging from lower-level chain/rule methods to higher-level globalnet methods. It seems the latter should ideally reside under the globalnet package and I'd imagine the implementations would boil down to lower-level primitives (ie chain/rule functions), in which case, we'd just need to abstract the primitives (a cursory look at a couple methods indicates that). Same with the MSS clamp methods. 

I suggest to start smaller with the lower-level primitives, ie first consider the `ipset` and `iptable` functionality. We've already abstracted `ipset` and `iptable` interfaces for unit testing. Can `nftables` also be abstracted behind those interfaces?

yboaron replied:
I created separate APIs for MssClamping and GlobalNetEgress functions. 
The plan is to add new commit to this PR with all the client changes

tpantelis replied:
FWIR, we were going to first add the chains and rules interface and the implementations (iptables and fake for now) and convert consumers (all with separate commits). Then we'd consider whether separate GlobalNet and MssClamping abstractions were needed or use the `ipset` and `packetfilter` interfaces directly.

yboaron committed:
Update clients to use Packetfilter

Signed-off-by: Yossi Boaron <yboaron@redhat.com>

tpantelis commented on pkg/packetfilter/packetfilter.go:
```diff
@@ -0,0 +1,273 @@
+	SnatCIDR string
+	DnatCIDR string
+
+	OutIntreface string
```
```suggestion
	OutInterface string
```

tpantelis commented on pkg/packetfilter/packetfilter.go:
```diff
@@ -0,0 +1,273 @@
+
+var (
+	logger                 = log.Logger{Logger: logf.Log.WithName("PacketFilter")}
+	NewFunc                func() (Base, error)
```
I assume this is for unit tests to hook in like the current `iptables` abstraction but we don't need this  - the fake implementation can register with `AddDriver`.

yboaron replied:
Thought to use it for unit testing hook, but yeah , I guess we can use AddDriver instead.

tpantelis commented on pkg/packetfilter/packetfilter.go:
```diff
@@ -0,0 +1,273 @@
+	selectedDriverIndex int
+)
+
+func AddDriver(name string, isDefault bool, createFunc DriverCreateFunc) {
```
What if multiple register as the default? Below it looks like it always takes the default, which I assume will be `iptables`. My understanding is that we'd dynamically determine which driver to use based on what's installed/supported, in which case the `DriverCreateFunc` would also return a bool indicating if it's supported (an err indicates it couldn't be determined). If both `iptables` and `nftables` can be installed simultaneously then we'd need a priority ranking to decide.

yboaron replied:
I thought each driver should be responsible for determining if it is supported and register only if it is supported. 

Submariner should use the same packetfilter driver that cni and kube-proxy use, from what I understand cni and kube-proxy will use either iptables or nftables, so only one of iptables,nftables should be registered as a driver.

But since we still don't know how cni and kube-proxy will migrate to nftables I added the isDefault param. maybe I should delete it ?

tpantelis replied:
> I thought each driver should be responsible for determining if it is supported and register only if it is supported.

yeah we could do it that way but that might imply that only one driver should ever register, otherwise we'd still have to reconcile multiple registrations. We could just have _main.go_ figure out which driver to use and change this to `SetDriver` or, better yet, pass in the driver instance to `New`. Then there's be no ambiguity.

tpantelis replied:
Actually we wouldn't even need the `New` function.

tpantelis commented on pkg/packetfilter/packetfilter.go:
```diff
@@ -0,0 +1,273 @@
+	}
+}
+
+type Adapter struct {
```
So following the current `iptables` et al abstractions, this would be named `Interface` and would contain higher-level helper/utility methods that wrap one or more `Base` methods. This avoids implementations duplicating code. So following the [iptables](https://github.com/submariner-io/submariner/blob/628f134dae339c9cbd94ce23f2ea1177ab9cfdc9/pkg/iptables/iptables.go#L41) abstraction, `CreateIPHookChainIfNotExists`, `CreateRegularChainIfNotExists` et al would be in this interface.

tpantelis commented on pkg/packetfilter/iptables/iptables.go:
```diff
@@ -0,0 +1,578 @@
+	chainHookToStr  = [packetfilter.ChainHookMAX]string{"PREROUTING", "INPUT", "FORWARD", "OUTPUT", "POSTROUTING"}
+	ruleActiontoStr = [packetfilter.RuleActionMAX]string{"", "ACCEPT", "TCPMSS", "MARK", "SNAT", "DNAT"}
+	logger          = log.Logger{Logger: logf.Log.WithName("IPTables")}
+	NewFunc         func() (packetfilter.Driver, error)
```
This isn't needed.

tpantelis commented on pkg/packetfilter/iptables/iptables.go:
```diff
@@ -0,0 +1,578 @@
+	markValue  string
+}
+
+type Adapter struct {
```
This doesn't need to be public. Also it's an implementation not an adapter so I suggest to name it `iptablesImpl`.

tpantelis commented:
Since the `packetfilter` package is intended to replace `iptables`, we had also discussed renaming/moving directories/files to maintain git history continuity. So first rename  the _iptables_ dir to _packetfilter_ then move _iptables.go_ under _packetfilter/iptables_ as a starting point.

sridhargaddam commented on pkg/globalnet/controllers/egress_pod_watcher.go:
```diff
@@ -95,8 +96,8 @@ func (w *egressPodWatcher) onCreateOrUpdate(obj runtime.Object, _ int) bool {
 
-	if err := w.namedIPSet.AddEntry(pod.Status.PodIP, true); err != nil {
-		logger.Errorf(err, "Error adding pod IP %q to IP set %q", pod.Status.PodIP, w.ipSetName)
+	if err := w.pfIface.AddEgressPodIP(w.gnHandle, pod.Status.PodIP); err != nil {
```
Suppose we have a use-case where we want all the pods from a certain namespace to use a specific egressIP, we create a namedIPSet which has a unique name created using the key (i.e., namespace/name) and and store the PodIPs that are part of that namespace. It is not clear if we are doing the same in the new code. Can you explain how we are handling it? Is gnHandle supposed to be a unique name derived based on the key?

sridhargaddam commented on pkg/globalnet/controllers/gateway_monitor.go:
```diff
@@ -452,94 +455,148 @@ func (g *gatewayMonitor) createGlobalNetMarkingChain() error {
+		TargetChain: constants.SmGlobalnetEgressChainForCluster,
+		Action:      packetfilter.RuleActionJump,
+	}
+	if err := g.pFilter.InsertUnique(packetfilter.TableTypeNAT, constants.SmGlobalnetEgressChain, 6, &ruleSpec); err != nil {
```
Just to confirm, are you planning to use the same type of chains/flow (as used in IPtables approach) even with Nftables?

yboaron replied:
Yep, for the regular/base chains and rules, 
IPHook chains should be used for ipstack hook chains.
As per tables not sure which tables we'll use for nftables (this is why we use tabletype in new API )

sridhargaddam commented on pkg/globalnet/controllers/packetfilter/iface.go:
```diff
@@ -0,0 +1,369 @@
+	return nil
+}
+
+func (i *pfilter) GetKubeProxyClusterIPServiceChainName(service *corev1.Service,
```
We no longer depend on the kubeproxy serviceChain for supporting GN. This API can be deleted.

sridhargaddam commented on pkg/packetfilter/packetfilter.go:
```diff
@@ -0,0 +1,252 @@
+
+type TableType uint32
+
+type GlobalEgressHandle string
```
Please add a note/help on what this handle points to.

sridhargaddam commented on pkg/packetfilter/packetfilter.go:
```diff
@@ -0,0 +1,252 @@
+
+func AddDriver(name string, createFunc DriverCreateFunc) {
+	if drivers[name] != nil {
+		logger.Warningf("Multiple drivers attempting to register with name %q", name)
```
I think we should treat this error as FATAL or return an appropriate error here.

tpantelis replied:
I think we should do away with adding drivers as I mentioned here https://github.com/submariner-io/submariner/pull/2835#discussion_r1432722928 and simply let the caller figure out which driver to use and pass it to `New`, `func New(d Driver) (Interface, error)`. Unit tests would pass in the fake `Driver`. This would greatly simplify things.

tpantelis commented on pkg/packetfilter/packetfilter.go:
```diff
@@ -0,0 +1,252 @@
+	AppendUnique(table TableType, chain string, ruleSpec *Rule) error
+}
+
+type MssClampInterface interface {
+	InitMultiClusterMssClamp() error
+	DelMultiClusterMssClamp() error
+	SetMultiClusterMssClamp(clampType MssClampType, mssValue int) error
+	AddSubMultiClusterMssClamp(localCIDRs []string, destCIDRs []string) error
+	DelSubMultiClusterMssClamp(localCIDRs []string, destCIDRs []string) error
+}
+type GlobalNetEgressInterface interface {
+	CreateGlobalEgressForPods(key, snatCIDR, markValue string) (GlobalEgressHandle, error)
+	DeleteGlobalEgressForPods(handle GlobalEgressHandle) error
+	CreateGlobalEgressForNamespace(namespace, snatCIDR, markValue string) (GlobalEgressHandle, error)
+	DeleteGlobalEgressForNamespace(handle GlobalEgressHandle) error
+	AddPodIPToGlobalEgress(handle GlobalEgressHandle, podIP string) error
+	RemovePodIPToGlobalEgress(handle GlobalEgressHandle, podIP string) error
+	UnInstallGlobalEgress() error
+}
```
FWIR, we last discussed not including these specific higher-level component APIs, ie `packetfilter` would simply be an iptables/nftables abstraction. If we need such higher-level abstractions, they should be implemented in the respective component packages. But let's first start with low-level iptables/nftables functionality.

tpantelis commented on pkg/packetfilter/iptables/iptables.go:
```diff
@@ -0,0 +1,617 @@
+	IPSetIface              ipset.Interface
+}
+
+type Basic interface {
```
We shouldn't define an interface here. This package should implement `packetfilter.Interface`.

tpantelis replied:
bump

tpantelis commented on pkg/packetfilter/packetfilter.go:
```diff
@@ -0,0 +1,252 @@
+	UnInstallGlobalEgress() error
+}
+
+type Driver interface {
```
We really only need one publicly consumed interface. In the existing iptables API, we define a `Basic` interface that is implemented by the iptables wrapper and the fake one. We also define the public `Interface` that defines higher-level helper methods that use the low-level `Basic` interface. I think we should follow that same paradigm here (perhaps renaming `Basic` to `Driver`), assuming similar higher-level  methods still apply, eg `CreateRegularChainIfNotExists`, `PrependUnique` etc. But if they don't apply then we would just instantiate the desired `Interface` implementation.

tpantelis commented on pkg/packetfilter/fake/iptables.go:
```diff
@@ -24,46 +24,84 @@ import (
+		tableChains: map[string]set.Set[string]{},
+	}
+
+	a := &iptables.IPTablesImpl{
```
This should be a fake implementation of the packetfilter.Interface and not implement/reference/use iptables, ie it should be a sibling of iptables.

tpantelis committed:
Rename iptables package to packetfilter

Signed-off-by: Tom Pantelis <tompantelis@gmail.com>

tpantelis commented:
I pushed https://github.com/submariner-io/submariner/pull/2871 based on my comments to serve as a starting point for the general structure. I left the packetfilter interface the same as it was.  I think the logical next step would be to generalize the interface as is done in this PR. Then I think it would be useful to implement the interface using nftables and see if any changes are needed. Eg, can `InsertUnique`, `PrependUnique`, `UpdateChainRules` etc be implemented in the adapter interface as they are now?

yboaron commented:
> I pushed #2871 based on my comments to serve as a starting point for the general structure. I left the packetfilter interface the same as it was. I think the logical next step would be to generalize the interface as is done in this PR. Then I think it would be useful to implement the interface using nftables and see if any changes are needed. Eg, can `InsertUnique`, `PrependUnique`, `UpdateChainRules` etc be implemented in the adapter interface as they are now?

The following came up during the high level design discussions:

1. current API should be updated to support table types (nftables not using default tables as iptables)and IPHook chains (nftables distinguish between regular chains and iphook chains).

2. Rules spec API , should we keep iptables flags rules api also for nftables configuration or use generic API ? 
 we thought that it makes sense to use generic API

3. ipset/nftables set scope,  on which scope sets should be handled ? 
the idea was to handle/hide sets under packetfilter umbrella, also packetfilter can expose high-level APIs for mssclamp and globalnetegress

I think that '1' is mandatory.

As per '3', I assume that we can implement the high level APIs at packetfilter adapter,  it will call chains/rules/set APIs of backend driver (I need to update my PR with this change), If we want to keep it on the clients side we should expose generic ipsets/nftables set API

Not 100% sure about high-level APIs support, but I think implementing it as packetfilter adapter makes sense.

tpantelis commented on pkg/packetfilter/fake/packetfilter.go:
```diff
@@ -0,0 +1,469 @@
+func New() *PacketFilter {
+	if driverInstance != nil {
+		return driverInstance
+	}
```
I don't think we should define a global instance. I did so initially in my PR but ran into issues due to the retained state (as I'm sure you did). You've added `ClearDriver` to alleviate that but that requires every consumer to explicitly call it. It's simpler and cleaner to have `New` return a new instance on each call. In my version I also call `SetNewDriverFn` for convenience.

```
func New() *PacketFilter {
	pf := &PacketFilter{
		chainRules:  map[string]set.Set[string]{},
		tableChains: map[string]set.Set[string]{},
	}

	packetfilter.SetNewDriverFn(func() (packetfilter.Driver, error) {
		return pf, nil
	})

	return pf
}
```

tpantelis commented:
> As per '3', I assume that we can implement the high level APIs at packetfilter adapter, it will call chains/rules/set APIs of backend driver (I need to update my PR with this change), If we want to keep it on the clients side we should expose generic ipsets/nftables set API
> 
> Not 100% sure about high-level APIs support, but I think implementing it as packetfilter adapter makes sense.

Would the ipsets API be implemented differently with nftables? If so then we would need some abstraction for ipsets. The only thing with the high level APIs you've defined is they are specific to the MTU and GlobalNet components and I think they really should reside with those components. Perhaps we could define those APIs more generically so they aren't specific to MSS clamp and GlobalNet? Even so, I suspect we may have a lot of duplicate code in the drivers (as is the case with the fake one in this PR). The other solution, as you mentioned, is to generalize the ipsets API. We could include that API in the packerfilter API (ie the `Driver` interface). It seems to me this would be better.

tpantelis commented on pkg/packetfilter/iptables/iptables.go:
```diff
@@ -0,0 +1,612 @@
+	DeleteChain(table, chain string) error
+}
+
+func NewDriver() (packetfilter.Driver, error) {
```
nit - "Driver" is redundant.

```suggestion
func New() (packetfilter.Driver, error) {
```

tpantelis commented on pkg/packetfilter/packetfilter.go:
```diff
@@ -0,0 +1,208 @@
+
+	// rules
+	Delete(table TableType, chain string, ruleSpec *Rule) error
+	InsertUnique(table TableType, chain string, position int, ruleSpec *Rule) error
+	PrependUnique(table TableType, chain string, ruleSpec *Rule) error
+	UpdateChainRules(table TableType, chain string, rules []*Rule) error
```
These methods were originally in the `Adapter`. The iptables lib defines `AppendUnique` so it wasn't in the `Adapter`. Do you know what nftables lib we would use and if it defines similar functions? If not, perhaps we could leave these in the `Adapter` for now? `UpdateChainRules` seems high-level and would probably belong in the `Adapter` to avoid duplication.

tpantelis replied:
@yboaron Can you please reply to this comment? Since we don't know yet what the nftables impl will look like, let's keep `InsertUnique`, `PrependUnique`, and `UpdateChainRules` as is in the `Adapter` so we don't have code duplicated between iptables and the fake impl (that was the purpose of the `Adapter`). We can adjust, if needed, when the nftables impl is introduced.

Also `Driver` and `Interface` are backwards, ie `Driver` is the low-level API that the adapter `Interface` builds on.  So `InsertUnique`, `PrependUnique`, and `UpdateChainRules` should be defined in `Interface` and the rest in `Driver`.

yboaron commented:
> > As per '3', I assume that we can implement the high level APIs at packetfilter adapter, it will call chains/rules/set APIs of backend driver (I need to update my PR with this change), If we want to keep it on the clients side we should expose generic ipsets/nftables set API
> > Not 100% sure about high-level APIs support, but I think implementing it as packetfilter adapter makes sense.
> 
> Would the ipsets API be implemented differently with nftables? If so then we would need some abstraction for ipsets. The only thing with the high level APIs you've defined is they are specific to the MTU and GlobalNet components and I think they really should reside with those components. Perhaps we could define those APIs more generically so they aren't specific to MSS clamp and GlobalNet? Even so, I suspect we may have a lot of duplicate code in the drivers (as is the case with the fake one in this PR). The other solution, as you mentioned, is to generalize the ipsets API. We could include that API in the packerfilter API (ie the `Driver` interface). It seems to me this would be better.

Yeah, we should have separate implementation for nftables sets.  
I'll check if we can have more generic APIs for mssclamp and gnegress

tpantelis commented:
> Yeah, we should have separate implementation for nftables sets. I'll check if we can have more generic APIs for mssclamp and gnegress

Let's first start with iptables and get that reviewed and merged. Then:

- I think it would be useful to implement the nftables version
- Generalize ipsets with nftables
- Consider higher-level APIs

yboaron commented:
> Let's first start with iptables and get that reviewed and merged. Then:
> 
> * I think it would be useful to implement the nftables version
> * Generalize ipsets with nftables
> * Consider higher-level APIs

I updated my PR to: 
-  support generalize named set 
-  move the higher level APIs from driver layers
- implement higher level APIs using basic APIs (chain/rules and namedset) at packetfilter layer

tpantelis commented:
> > Let's first start with iptables and get that reviewed and merged. Then:
> > 
> > * I think it would be useful to implement the nftables version
> > * Generalize ipsets with nftables
> > * Consider higher-level APIs
> 
> I updated my PR to:
> 
> * support generalize named set
> * move the higher level APIs from driver layers
> * implement higher level APIs using basic APIs (chain/rules and namedset) at packetfilter layer

I'd like to do these in separate PRs. The steps I outlined above would be separate PRs. I'd like to first get the existing iptables interface **as is** including the adapter (with InsertUnique etc) moved to packetfilter and generalized. Basically that's my PR with your generalizations. Then get that merged. Then  implement the nftables version possibly revising the adapter methods and get that reviewed/merged. Then generalize ipsets with iptables and nftables impls in a separate PR. Hopefully, at that point, everything will work with nftables. If not or if we think we need higher-level APIs then look at that.

Does this make sense?

yboaron commented:
> I'd like to do these in separate PRs. The steps I outlined above would be separate PRs. I'd like to first get the existing iptables interface **as is** including the adapter (with InsertUnique etc) moved to packetfilter and generalized. Basically that's my PR with your generalizations. Then get that merged. Then implement the nftables version possibly revising the adapter methods and get that reviewed/merged. Then generalize ipsets with iptables and nftables impls in a separate PR. Hopefully, at that point, everything will work with nftables. If not or if we think we need higher-level APIs then look at that.
> 
> Does this make sense?


The existing iptables API can't be used for nftables (IPHook chain and table types are missing), so first we need to update the API to support also nftables.
  
Also for implementing and testing nftables we depend on CNI and kube-proxy nftables implementation,  since no tables and iphook chains  are created by default for nftables , we need to see what tables/iphook chains the CNI and kube-proxy are using. and decide if we want to use these tables/chains or create our own.

So I would say that the tasks are : 

1. We first need to do all the "preparations" for nftables/generalization work ,  for this we need to generalize  iptables API and add support for tabletype and iphook chain.
Basically it is APIs under [packetfilter.Interface ](https://github.com/submariner-io/submariner/pull/2835/files#diff-8ce084ff520c3b3607c1cbcaf08a5490108365f84e3e50646549907eb72081dbR172)

task 1 includes: 
1.1 Generalize named set API
1.2 Of course we need to Implement the iptables backend driver.
1.3 Update the clients to use the new API 

2. As part of the nftables work, there was a proposal to provide high level APIs in the context of packetfilter for GNEgress and MssClamp, this should simplify and separate things at the clients side (especially for GN egress controllers) .

As you mentioned  we can implement these high-level api also at the clients side, but either case I think we should add these high level APIs. 

These 2 interfaces :
[mss-clamp](https://github.com/submariner-io/submariner/pull/2835/files#diff-8ce084ff520c3b3607c1cbcaf08a5490108365f84e3e50646549907eb72081dbR192)

[GN egress](https://github.com/submariner-io/submariner/pull/2835/files#diff-8ce084ff520c3b3607c1cbcaf08a5490108365f84e3e50646549907eb72081dbR199)


3. Add nftables backend driver
We need to support nftables by ~mid 2025,  so we still have time here.
Also for implementing nftables we depend on CNI and kube-proxy nftables implementation (https://github.com/kubernetes/kubernetes/issues/122572) 

My current PR , covers 1 and 2.

I think we should: 
A. first push 1 and 2  in 0.17 , covered by this PR.
B. Have some time for testing Submariner with packetfilter/iptables mode in 0.17/dev.
C. In next SM release (0.18?) we can start working on nftables backend driver, and run some testing with kube-proxy and CNI run in nftables mode.

tpantelis commented:
> > I'd like to do these in separate PRs. The steps I outlined above would be separate PRs. I'd like to first get the existing iptables interface **as is** including the adapter (with InsertUnique etc) moved to packetfilter and generalized. Basically that's my PR with your generalizations. Then get that merged. Then implement the nftables version possibly revising the adapter methods and get that reviewed/merged. Then generalize ipsets with iptables and nftables impls in a separate PR. Hopefully, at that point, everything will work with nftables. If not or if we think we need higher-level APIs then look at that.
> > Does this make sense?
> 
> The existing iptables API can't be used for nftables (IPHook chain and table types are missing), so first we need to update the API to support also nftables.

I know  - I stated that above: _I'd like to first get the existing iptables interface **as is** including the adapter (with InsertUnique etc) moved to packetfilter and **generalized_**

That should be the first PR.
 
> 1. We first need to do all the "preparations" for nftables/generalization work ,  for this we need to generalize  iptables API and add support for tabletype and iphook chain.

Which is this PR.

> task 1 includes: 1.1 Generalize named set API 1.2 Of course we need to Implement the iptables backend driver. 1.3 Update the clients to use the new API

In a separate PR after this. Let's do this in stages to keep the PRs smaller and easier to review.

> 
> 2. As part of the nftables work, there was a proposal to provide high level APIs in the context of packetfilter for GNEgress and MssClamp, this should simplify and separate things at the clients side (especially for GN egress controllers) .
> 
> As you mentioned we can implement these high-level api also at the clients side, but either case I think we should add these high level APIs.

I'm not convinced we need these. They're specific to certain components with basically one consumer so not generally re-usable, at least from what I see right now. Maybe the introduction of nftables will change that and require more generalization but let's see how that shakes out. I'd like to see the nftables implementation first.

> 3. Add nftables backend driver
>    We need to support nftables by ~mid 2025,  so we still have time here.
>    Also for implementing nftables we depend on CNI and kube-proxy nftables implementation ([nftables kube-proxy TODO kubernetes/kubernetes#122572](https://github.com/kubernetes/kubernetes/issues/122572))

OK I thought nftables was for this (0.17) release.

> I think we should: A. first push 1 and 2 in 0.17 , covered by this PR. 

As mentioned earlier, I think we should hold off on 2 for now.

yboaron commented:
> In a separate PR after this. Let's do this in stages to keep the PRs smaller and easier to review.

Hmmm, let's have the generalization API (including the single API for namedset) in this first PR, I'll drop 2.

As per InsertUnique, PrependUnique, and UpdateChainRules, yep it makes sense to keep the logic in the adapter, but that means we should do some adjustments to List API and maybe add Contains API, do you think we can have this optimization in a follow-up  PR ?

tpantelis commented:
> > In a separate PR after this. Let's do this in stages to keep the PRs smaller and easier to review.
> 
> Hmmm, let's have the generalization API (including the single API for namedset) in this first PR, I'll drop 2.

I'd still prefer to add the IP set generalization in a separate PR to make it a bit easier to review. The PR is already big with the iptables stuff.

> 
> As per InsertUnique, PrependUnique, and UpdateChainRules, yep it makes sense to keep the logic in the adapter, but that means we should do some adjustments to List API and maybe add Contains API, do you think we can have this optimization in a follow-up PR ?

Well that's the way the current API is defined so I think it makes sense to keep it as and adjust the generalization accordingly.  We can just keep the interfaces as is. The adapter methods use `List`, `Insert`, and `Delete`. The latter two are already generalized or simple to do so. I guess `List` would return `Rule` objects so we'd have to parse out the rule string which should be pretty straightforward.

sridhargaddam commented:
@yboaron not sure if you saw it already. Kube proxy seems to be supporting nftables backend now as an alpha feature starting with 1.29 version - https://kubernetes.io/docs/reference/networking/virtual-ips/#proxy-mode-nftables

So, we can use a corresponding KIND setup to verify any PRs.

kubeproxy PR - https://github.com/kubernetes/kubernetes/pull/121046

sridhargaddam commented on pkg/packetfilter/fake/packetfilter.go:
```diff
@@ -0,0 +1,369 @@
+	return nil
+}
+
+func (i *PacketFilter) CreateRegularChainIfNotExists(table packetfilter.TableType, chain *packetfilter.ChainRegular) error {
```
What is the purpose of having "Regular" in the API - CreateRegularChainIfNotExists?

yboaron commented:
> @yboaron not sure if you saw it already. Kube proxy seems to be supporting nftables backend now as an alpha feature starting with 1.29 version - https://kubernetes.io/docs/reference/networking/virtual-ips/#proxy-mode-nftables
> 
> So, we can use a corresponding KIND setup to verify any PRs.
> 
> kubeproxy PR - [kubernetes/kubernetes#121046](https://github.com/kubernetes/kubernetes/pull/121046)

Yep, saw it, they are using[ separate table ]
(https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/nftables/proxier.go#L62) for their configurations. 

I assume that after the code implementing the transition to packetfilter gets merged, we can use this KIND setup with nftable for testing of our nftables driver. 

I guess we need also the CNI (KindNet) switched to nftables , for full e2e testing on Kind, right ?

tpantelis reviewed (changes requested):
Your first commit moves _pkg/iptables/iptables.go_  so we retain Git history there but we've lost the history for _pkg/iptables/adapter.go_ and _pkg/iptables/fake/iptables.go_. I think we should first move/rename the entire directory as I did in my PR. So I suggest to start with the 3 commits from my PR and then refactor the packetfilter APIs with your generalization and then adjust consumers in 2 separate commits. I think that should produce a pretty clean refactor and reasonable to review.

sridhargaddam commented:
> I guess we need also the CNI (KindNet) switched to nftables , for full e2e testing on Kind, right ?

Yes. Kindnet programs certain IPtable rules mainly to perform MASQ for outbound traffic so that the pods can reach internet. As you are familiar, `iptables-nft` was created to facilitiate smooth migration to `nftables` and pretty much all the CNIs (along with kubeproxy) use `iptables-nft` under the hood today. 
So, if the nftables handler in Submariner uses `nft` while the CNI uses `iptables-nft` AFAICT it should still continue to work, assuming we take care of the priority to enforce the order in which the rules have to be applied.

yboaron committed:
Add support for packetfilter

To allow a smooth transition to nftables a new packetfilter component
is introduced, packetfilter provides an API for creating chains and
rules in a generic way and should include pluggable drivers for
iptables and nftables.

Packetfilter also includes a static function that defines the driver
to use.

This commit includes:
- Refactoring the current iptables code to be used as pluggable driver.
- Fake driver implementation for unit tests.

Co-authored-by: Tom Pantelis <tompantelis@gmail.com>
Signed-off-by: Yossi Boaron <yboaron@redhat.com>

yboaron committed:
Rename kubeproxy/kp_iptables.go and kubeproxy/iptables_iface.go

Signed-off-by: Yossi Boaron <yboaron@redhat.com>

yboaron commented:
With the latest version I pushed, we have the history pointing to the prev files for the relevant files [1] , 
do we need something else ? 

[1]
git log --pretty=oneline  --follow pkg/packetfilter/adapter.go
git log --pretty=oneline  --follow pkg/packetfilter/packetfilter.go
git log --pretty=oneline  --follow pkg/packetfilter/fake/packetfilter.go
git log --pretty=oneline --follow pkg/routeagent_driver/handlers/kubeproxy/packetfilter_iface.go
git log --pretty=oneline --follow pkg/routeagent_driver/handlers/kubeproxy/kp_packetfilter.go

tpantelis commented on pkg/packetfilter/packetfilter.go:
```diff
@@ -0,0 +1,192 @@
+	List(table TableType, chain string) ([]string, error)
+	DeleteRuleSpec(table TableType, chain string, rulespec ...string) error
+	Append(table TableType, chain string, rulespec ...string) error
+	Insert(table TableType, chain string, pos int, rulespec ...string) error
```
So we have some rule methods that take a generalized `Rule` and some that take a rule string. This seems inconsistent. Also the string form is implementation-specific and I don't think we should expose via the generalized interface, ie all methods should reference `Rule`.  I assume you exposed the string form for the `List` method. However, for itpables, it seems if we can convert `Rule` to the itpables string form we should be able to convert the string back to a `Rule`. So `List` would be `List(table TableType, chain string) ([]Rule, error)`. In the `Adapter`, we should be able to compare `Rule` instances. That would also obviate defining 2 driver interfaces to avoid exposing the stringified versions. The interface would be:

```
type Driver interface {
	// Chains
	ChainExists(table TableType, chain string) (bool, error)
	CreateIPHookChainIfNotExists(chain *ChainIPHook) error
	CreateChainIfNotExists(table TableType, chain *Chain) error
	DeleteIPHookChain(chain *ChainIPHook) error
	DeleteRegularChain(table TableType, chain string) error
	ClearChain(table TableType, chain string) error

	// rules
	Delete(table TableType, chain string, ruleSpec *Rule) error
	AppendUnique(table TableType, chain string, ruleSpec *Rule) error
        Append(table TableType, chain string, ruleSpec *Rule) error
	Insert(table TableType, chain string, pos int, ruleSpec *Rule)
        List(table TableType, chain string) ([]Rule, error)
}
```

tpantelis replied:
I wrote a `FromRuleSpec` function in [this commit](https://github.com/tpantelis/submariner/pull/84/commits/b144b9040e3d34f7da3793b19dfdd01319687290) with unit tests. The `Rule`s tested are taken from consumer code to cover all the rule actions.

yboaron replied:
I copied your FromRuleSpec function (added you as a co-author )  , thanks.

yboaron commented:
> Yes. Kindnet programs certain IPtable rules mainly to perform MASQ for outbound traffic so that the pods can reach internet. 

My Q was more about the nftables tables Submariner should use/create, 
we know that Kube-proxy uses its own table, and a CNI that uses iptables-nft use the default tables and chains similar to iptables while other CNIs that use nftables may create their own table/s.

I guess we can't assume that default tables/chains (iptables-nft) will always be there, WDYT?  meaning we should create our own table.

tpantelis commented on pkg/packetfilter/fake/packetfilter.go:
```diff
@@ -0,0 +1,359 @@
+	"k8s.io/utils/set"
+)
+
+const (
+	remoteCIDRIPSet                    = "SUBMARINER-REMOTECIDRS"
+	localCIDRIPSet                     = "SUBMARINER-LOCALCIDRS"
+	smPostRoutingChain                 = "SUBMARINER-POSTROUTING"
+	smGlobalnetEgressChainForPods      = "SM-GN-EGRESS-PODS"
+	smGlobalnetEgressChainForNamespace = "SM-GN-EGRESS-NS"
+	gnIPSetPrefix                      = "SM-GN-"
+	DriverName                         = "fake"
+)
```
These are unused and should not be defined here since they're consumer-specific.

```suggestion
```

tpantelis commented on pkg/packetfilter/fake/packetfilter.go:
```diff
@@ -0,0 +1,359 @@
+		return errors.Wrapf(err, "error creating IP tables %s:%s chain", iphookChainTypeToStr[chain.Type], chain.Name)
+	}
+
+	jumpRule := chain.JumpRule
+	if jumpRule == nil {
+		jumpRule = &packetfilter.Rule{
+			TargetChain: chain.Name,
+			Action:      packetfilter.RuleActionJump,
+		}
+	}
+
+	if chain.Priority == packetfilter.ChainPriorityFirst {
+		rulespec, err := i.RuleToRuleSpec(jumpRule)
+		if err != nil {
+			return errors.Wrap(err, "error translating jumpRule")
+		}
+
+		if err := i.Insert(iphookChainTypeToTableType[chain.Type], chainHookToStr[chain.Hook], 1, rulespec...); err != nil {
+			return errors.Wrap(err, "error PrependUnique rule")
+		}
+	} else {
+		if err := i.AppendUnique(iphookChainTypeToTableType[chain.Type], chainHookToStr[chain.Hook], jumpRule); err != nil {
+			return errors.Wrap(err, "error AppendUnique rule")
+		}
+	}
```
Is is necessary to add a jump rule in the fake implementation? Isn't this iptables-specific? In `AwaitChain`, we only verify the chain is present in `tableChains`.

yboaron replied:
Yeah, I guess we can drop this code. I'll update it.

tpantelis commented:
> With the latest version I pushed, we have the history pointing to the prev files for the relevant files [1] , do we need something else ?

That looks good.

tpantelis commented on pkg/packetfilter/fake/packetfilter.go:
```diff
@@ -0,0 +1,359 @@
+}
+
+func (i *PacketFilter) AppendUnique(table packetfilter.TableType, chain string, ruleSpec *packetfilter.Rule) error {
+	ruleSpecStr, err := iptables.RuleToRuleSpecFn(ruleSpec)
```
We shouldn't have a dependency on iptables. We can simply stringify the `Rule`: `fmt.Sprintf("%+v", ruleSpec)`.

tpantelis commented on pkg/packetfilter/fake/packetfilter.go:
```diff
@@ -0,0 +1,359 @@
+	failOnDeleteRuleMatchers []interface{}
+}
+
+var (
+	tableTypeToStr             = [packetfilter.TableTypeMAX]string{"filter", "mangle", "nat"}
+	iphookChainTypeToStr       = [packetfilter.ChainTypeMAX]string{"filter", "mangle", "nat"}
+	iphookChainTypeToTableType = [packetfilter.ChainTypeMAX]packetfilter.TableType{
+		packetfilter.TableTypeFilter,
+		packetfilter.TableTypeRoute,
+		packetfilter.TableTypeNAT,
+	}
+	chainHookToStr = [packetfilter.ChainHookMAX]string{"PREROUTING", "INPUT", "FORWARD", "OUTPUT", "POSTROUTING"}
+)
```
These are iptables-specific. For the fake impl, we can use `packetfilter.TableType` directly instead of converting to a string. So `tableChains` can be defined as `map[packetfilter.TableType]set.Set[string]`. The key for `chainRules` combines table and chain which would become `chainRules[fmt.Sprintf("%v/%s", table, chain)]`

sridhargaddam commented:
> My Q was more about the nftables tables Submariner should use/create, we know that Kube-proxy uses its own table, and a CNI that uses iptables-nft use the default tables and chains similar to iptables while other CNIs that use nftables may create their own table/s.
> 
> I guess we can't assume that default tables/chains (iptables-nft) will always be there, WDYT? meaning we should create our own table.

AFAICT, if some component is using "iptables" (with `iptables-nft` under the hood) it will anyway use the default tables (like `nat`, `filter`, `mangle` etc) and it can create it own chains (like KUBE-PREROUTING etc) with a rule in the default chain (like PREROUTING) to forward packets to its custom chain (KUBE-PREROUTING). While using iptables they will not be creating any custom tables.

However, when a component has fully migrated to support `nft`, it can decide to create additional tables within each family and these tables can be given custom names. The components can then configure higher priority to those tables so that it gets priority over the default tables. 

Personally, I feel in `nftables` implementation, we can create our own tables with Submariner specific chains/rules and use a higher priority.

tpantelis commented on pkg/packetfilter/iptables/iptables.go:
```diff
@@ -0,0 +1,463 @@
+)
+
+const (
+	remoteCIDRIPSet                    = "SUBMARINER-REMOTECIDRS"
+	localCIDRIPSet                     = "SUBMARINER-LOCALCIDRS"
+	smPostRoutingChain                 = "SUBMARINER-POSTROUTING"
+	smGlobalnetEgressChainForPods      = "SM-GN-EGRESS-PODS"
+	smGlobalnetEgressChainForNamespace = "SM-GN-EGRESS-NS"
+	gnIPSetPrefix                      = "SM-GN-"
```
These are unused.

```suggestion
```

tpantelis commented on pkg/packetfilter/iptables/iptables.go:
```diff
@@ -0,0 +1,463 @@
+	smGlobalnetEgressChainForPods      = "SM-GN-EGRESS-PODS"
+	smGlobalnetEgressChainForNamespace = "SM-GN-EGRESS-NS"
+	gnIPSetPrefix                      = "SM-GN-"
+	DriverName                         = "IPTables"
```
This is a remnant from the initial implementation. It's now only used on L51 so can be inlined there.

tpantelis commented on pkg/packetfilter/iptables/iptables.go:
```diff
@@ -0,0 +1,463 @@
+
+	for _, existingRule := range existingRules {
+		ruleSpec := strings.Split(existingRule, " ")
+		if ruleSpec[0] == "-A" {
+			ruleSpec = ruleSpec[2:] // remove "-A", "$chain"
+		}
```
Why is this done? If it's necessary, can you add a comment as to why?

yboaron replied:
copied it from [current iptables code](https://github.com/submariner-io/submariner/blob/devel/pkg/iptables/adapter.go#L111), ipt.List function returns rules list in format similar to [1] , as specified in the comment just remove '-A chainmae' from rules list. 


[1] 
`# iptables -t nat -S SUBMARINER-POSTROUTING 

-A SUBMARINER-POSTROUTING -s 240.0.0.0/8 -o vx-submariner -j SNAT --to-source 10.131.1.1
-A SUBMARINER-POSTROUTING -s 10.131.0.0/16 -d 242.0.0.0/16 -j ACCEPT
-A SUBMARINER-POSTROUTING -s 242.0.0.0/16 -d 10.131.0.0/16 -j ACCEPT
`

tpantelis replied:
ok

tpantelis commented on pkg/packetfilter/iptables/iptables.go:
```diff
@@ -0,0 +1,463 @@
+	return ruleSpec, nil
+}
+
+func fromRuleSpec(spec []string) *packetfilter.Rule {
```
Can you add the unit tests I wrote from https://github.com/tpantelis/submariner/pull/84/files#diff-a7bc1ab2973d80adbeacece70b4b58932ad76cf96581e19ca7cc07463b395e4d?

tpantelis commented on pkg/packetfilter/packetfilter.go:
```diff
@@ -0,0 +1,186 @@
+
+type TableType uint32
+
+type GlobalEgressHandle string
```
This is unused now.

```suggestion
```

tpantelis commented on pkg/globalnet/controllers/packetfilter/iface.go:
```diff
@@ -0,0 +1,319 @@
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type GlobalEgressHandle string
```
This is unused.

```suggestion
```

tpantelis reviewed (changes requested):
Overall looks good, just some min or things.

Please update the summary for https://github.com/submariner-io/submariner/pull/2835/commits/65c28b288d65c9bbc4e5da5c20498ebd0345602c to reflect that we didn't add general MSS clamping and GN APIs.  Also remove "_1. Packetfilter component, auto-detecting hardcoded to iptables._".

Similarly please update the PR summary.

yboaron committed:
Merge branch 'devel' into packetfilter

tpantelis committed:
Merge branch 'devel' into packetfilter
